### PR TITLE
Add batch norm grad support

### LINKED ISSUE #5334: Add support for batch_norm_grad
We need support for stablehlo.batch_norm_grad, which appears in bwd passes.
Reference: https://openxla.org/stablehlo/spec#batch_norm_grad

### PR DESCRIPTION
**Pull request by pglusacTT**
### Ticket
Closes #5334

### Problem description
We do not have support for [stablehlo.batch_norm_grad](https://openxla.org/stablehlo/spec#batch_norm_grad) which is invoked during traning.

### What's changed
- Added batch norm grad conversion pattern
- Updated batch norm training to use momentum = 1 and updated initialization of mean and variance
- Skipped consteval of values that have memory effects

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Transforms/ConstEvalHoist.cpp [hunk 1]
```diff
@@ -190,6 +190,27 @@ class ConstEvalAnalyze {
       }
     }
 
+    // Check if any of my results are written to, not safe for consteval
+    for (auto result : op->getResults()) {
+      // check all users of this result
+      for (auto user : result.getUsers()) {
```
Comment (vmilosevic) [inline lib/Transforms/ConstEvalHoist.cpp]: :warning: **llvm\-qualified\-auto** :warning:
`` auto user `` can be declared as `` auto *user ``
```suggestion
      for (auto *user : result.getUsers()) {
```

**Review by vmilosevic — COMMENTED** [review]
:warning: `Clang-Tidy` found issue(s) with the introduced code (1/1)

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 2]
```diff
@@ -772,11 +772,17 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
-    FloatAttr momentumAttr = rewriter.getF32FloatAttr(0.1f);
+    FloatAttr momentumAttr = rewriter.getF32FloatAttr(1.0f);
 
-    // Create empty tensors for running mean and variance
-    auto runningMean = rewriter.create<ttir::EmptyOp>(loc, meanType);
-    auto runningVariance = rewriter.create<ttir::EmptyOp>(loc, varianceType);
+    auto meanShapeAttr = rewriter.getDenseI32ArrayAttr(
+        llvm::SmallVector<int32_t>(meanType.getShape()));
+    auto varianceShapeAttr = rewriter.getDenseI32ArrayAttr(
+        llvm::SmallVector<int32_t>(varianceType.getShape()));
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Nit: you can do `llvm::to_vector_of<int32_t>(x)` without creation of attributes.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 3]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
 };
 } // namespace
 
+namespace {
+
+// Verify that scale, mean, variance parameters are 1D tensors with correct size
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Nit: end comment with the dot (in other places as well)

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 4]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
+static LogicalResult checkBatchNormGradParametersLegality(
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Nit: function is already in anon ns, no need for static. It might make more sense to create is a a private member function of `StableHLOToBatchNormGradOpConversionPattern` if that is going to be it's only user.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 5]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
+    mlir::stablehlo::BatchNormGradOp &srcOp,
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Nit: remove ref

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 6]
```diff
@@ -772,11 +772,17 @@ class StableHLOToBatchNormTrainingOpConversionPattern
         mlir::IntegerAttr::get(integerType, srcOp.getFeatureIndex());
 
     // Default momentum for batch norm training
-    FloatAttr momentumAttr = rewriter.getF32FloatAttr(0.1f);
+    FloatAttr momentumAttr = rewriter.getF32FloatAttr(1.0f);
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:775]: Why has this changed?

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 7]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
+    // Get the input types and shapes
+    auto operandType =
+        mlir::cast<RankedTensorType>(adaptor.getOperand().getType());
+    auto scaleType = mlir::cast<RankedTensorType>(adaptor.getScale().getType());
+    auto gradOutputType =
+        mlir::cast<RankedTensorType>(adaptor.getGradOutput().getType());
+
+    // Get output types
+    auto gradOperandType = mlir::cast<RankedTensorType>(
+        getTypeConverter()->convertType(srcOp.getResult(0).getType()));
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Prefer named over generic accessors (i.e. `srcOp.getGradOperand()`). Similar for the `gradScale` and `gradOffset`.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 8]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
+
+    // Get feature index and epsilon
+    int64_t featureIndex = srcOp.getFeatureIndex();
+    FloatAttr epsilonAttr = srcOp.getEpsilonAttr();
+    float epsilon = epsilonAttr.getValueAsDouble();
+
+    // Create reduction dimensions (all dims except feature_index)
+    int64_t rank = operandType.getRank();
+    SmallVector<int32_t> reductionDims(rank);
+    std::iota(reductionDims.begin(), reductionDims.end(), 0);
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Nit:
```suggestion
    SmallVector<int32_t> reductionDims = llvm::to_vector(llvm::seq<int32_t>(rank));
```

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 9]
```diff
@@ -795,6 +801,301 @@ class StableHLOToBatchNormTrainingOpConversionPattern
...
+        rewriter.getFloatAttr(operandType.getElementType(), epsilon));
+    auto epsilonConstant =
+        rewriter.create<ttir::ConstantOp>(loc, scalarType, epsilonDenseAttr);
+    Value epsilonBcast = broadcastFeatureToShape(rewriter, loc, epsilonConstant,
+                                                 operandType, featureIndex);
+
+    // centered_operand = operand - mean
+    auto centeredOperandEmpty =
+        rewriter.create<ttir::EmptyOp>(loc, operandType);
+    auto centeredOperand = rewriter.create<ttir::SubtractOp>(
```
Comment (azecevic000) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: You can use `ttir::utils::createDPSOp/replaceOpWithDPSOp` for automatic `EmptyOp` creation.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 6 above)
Comment (pglusacTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:775]: The 0.1 value was set because it is the default for torch, however that is not what the default value should be according to stablehlo specs. I caught this while debugging batch norm grad pcc.
